### PR TITLE
Diffie-Hellman tuples support in sigma protocol

### LINKED ISSUE #247: Sigma protocol implementation for DHT

### PR DESCRIPTION
**Pull request by greenhat**
Close #247

This PR adds support for Diffie-Hellman tuple in sigma protocol.
Prover messages are generated in `dht_protocol` module.
Besides existing `ProveDhTuple` (see `private_input` module) added `DhTupleProverInput` that holds both a secret and a tuple.
DH tuple added to sigma trees (see `Unproven...` and `Unchecked...` hierarchies) and handled in relevant steps in the prover (see `prover` module) and verifier (see `sig_serializer` and `verifier` module).
Tests are added in `prover`, `verifier` modules along with test vectors (shared with sigmastate) in `signing_spec_tests` test module. PBT tests for sigma conjectures are using `PrivateInput` generator that produces either descrete log input or DH tuple.
Corresponding test vectors in sigmastate are added in https://github.com/ScorexFoundation/sigmastate-interpreter/pull/729

### CODE REVIEW

File: ergotree-interpreter/src/sigma_protocol/private_input.rs [hunk 1]
```diff
@@ -76,13 +78,57 @@ impl From<Scalar> for DlogProverInput {
...
+    /// Diffie-Hellman tuple's secret
+    pub w: Scalar,
+    /// Diffie-Hellman tuple
+    pub common_input: ProveDhTuple,
+}
+
+impl DhTupleProverInput {
+    /// Create random secret and Diffie-Hellman tuple
+    #[allow(clippy::many_single_char_names)]
+    pub fn random() -> Self {
```
Comment (kushti) [inline ergotree-interpreter/src/sigma_protocol/private_input.rs:96]: Is it used in tests only? If yes, better to specify it in the comments (or move the function into tests?)

File: ergotree-interpreter/src/sigma_protocol/dht_protocol.rs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        let v_to_minus_e = dlog_group::exponentiate(&public_input.v, &minus_e);
+        let a = g_to_z.mul(&u_to_minus_e);
+        let b = h_to_z.mul(&v_to_minus_e);
+        (
+            FirstDhTupleProverMessage::new(a, b),
+            SecondDhTupleProverMessage { z },
+        )
+    }
+
+    pub(crate) fn first_message(
```
Comment (kushti) [inline ergotree-interpreter/src/sigma_protocol/dht_protocol.rs:73]: Please add a function description.

File: ergotree-interpreter/src/sigma_protocol/dht_protocol.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        public_input: &ProveDhTuple,
+    ) -> (Scalar, FirstDhTupleProverMessage) {
+        let r = dlog_group::random_scalar_in_group_range();
+        let a = dlog_group::exponentiate(&public_input.g, &r);
+        let b = dlog_group::exponentiate(&public_input.h, &r);
+        (r, FirstDhTupleProverMessage::new(a, b))
+    }
+
+    pub(crate) fn second_message(
```
Comment (kushti) [inline ergotree-interpreter/src/sigma_protocol/dht_protocol.rs:82]: Please add a function description.

File: ergotree-interpreter/src/sigma_protocol/dht_protocol.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    use std::ops::Mul;
+
+    use super::*;
+    use crate::sigma_protocol::private_input::DhTupleProverInput;
+    use crate::sigma_protocol::Challenge;
+    use ergotree_ir::sigma_protocol::dlog_group;
+    use ergotree_ir::sigma_protocol::sigma_boolean::ProveDhTuple;
+    use k256::Scalar;
+
+    pub(crate) fn simulate(
```
Comment (kushti) [inline ergotree-interpreter/src/sigma_protocol/dht_protocol.rs:51]: Please add a function description.

**Review by kushti — APPROVED** [review]
LGTM but please check comments

File: ergotree-interpreter/src/sigma_protocol/private_input.rs (hunk 1 above)
Comment (greenhat) [reply ergotree-interpreter/src/sigma_protocol/private_input.rs:96]: Yes. I moved it to tests. 
There is also `DlogProverInput::random()` which is used in tests but 
it's public and exposed in WASM layer(JS/TS). Do you prefer to remove `DlogProverInput::random()` from all public APIs as well?

File: ergotree-interpreter/src/sigma_protocol/dht_protocol.rs (hunk 2 above)
Comment (greenhat) [reply ergotree-interpreter/src/sigma_protocol/dht_protocol.rs:73]: Done.

File: ergotree-interpreter/src/sigma_protocol/dht_protocol.rs (hunk 3 above)
Comment (greenhat) [reply ergotree-interpreter/src/sigma_protocol/dht_protocol.rs:82]: Done.

File: ergotree-interpreter/src/sigma_protocol/dht_protocol.rs (hunk 4 above)
Comment (greenhat) [reply ergotree-interpreter/src/sigma_protocol/dht_protocol.rs:51]: Done.

File: ergotree-interpreter/src/sigma_protocol/private_input.rs (hunk 1 above)
Comment (kushti) [reply ergotree-interpreter/src/sigma_protocol/private_input.rs]: ah then it is reasonable to have both random() public and exposed. Please check that randomness is of cryptographic quality.

Comment (greenhat) [reply ergotree-interpreter/src/sigma_protocol/private_input.rs]: I switched to `ThreadRng` PRNG (more cryptographically secure than `OsRng`) and restored public `DhTupleProverInput::random()`
[`ThreadRng`](https://rust-random.github.io/rand/rand/rngs/struct.ThreadRng.html) is [`StgRng`](https://rust-random.github.io/rand/rand/rngs/struct.StdRng.html) (ChaCha block cipher with 12 as of July 2021) seeded from [`OsRng`](https://rust-random.github.io/rand/rand/rngs/struct.OsRng.html) which is a random number generator that retrieves randomness from the operating system.

**Pull request merged**